kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,624 @@
+            self.runtime.block_on(async {
+                self.store
+                    .put(&path, data.into())
+                    .await
+                    .expect("put log file in store");
+            });
+        }
+        pub(crate) fn table_root(&self) -> &Path {
+            self.dir.path()
+        }
+    }
+    fn get_init_commit() -> Vec<Action> {
+        let schema = StructType::new([
+            StructField::new("id", DataType::LONG, true),
+            StructField::new("value", DataType::STRING, true),
+        ]);
+        let schema_string = serde_json::to_string(&schema).unwrap();
+        vec![
+            Metadata {
+                schema_string,
+                configuration: HashMap::from([
+                    ("enableChangeDataFeed".to_string(), "true".to_string()),
+                    ("enableDeletionVectors".to_string(), "true".to_string()),
+                ]),
+                ..Default::default()
+            }
+            .into(),
+            Protocol::try_new(3, 7, Some(["deletionVectors"]), Some(["deletionVectors"]))
+                .unwrap()
+                .into(),
+        ]
+    }
+
+    fn get_segment(
+        engine: &dyn Engine,
+        path: &Path,
+        start_version: Version,
+        end_version: impl Into<Option<Version>>,
+    ) -> DeltaResult<Vec<ParsedLogPath>> {
+        let table_root = url::Url::from_directory_path(path).unwrap();
+        let log_root = table_root.join("_delta_log/")?;
+        let log_segment = LogSegment::for_table_changes(
+            engine.get_file_system_client().as_ref(),
+            log_root,
+            start_version,
+            end_version,
+        )?;
+        Ok(log_segment.ascending_commit_files)
+    }
+
+    fn get_commit_log_scanner(engine: &dyn Engine, commit: ParsedLogPath) -> LogReplayScanner {
+        let expression_evaluator = engine.get_expression_handler().get_evaluator(
+            get_log_add_schema().clone(),
+            get_add_transform_expr(),
+            SCAN_ROW_DATATYPE.clone(),
+        );
+        LogReplayScanner::new(
+            commit,
+            engine.get_json_handler(),
+            None,
+            expression_evaluator.clone(),
+        )
+    }
+
+    #[test]
+    fn metadata_protocol() {
+        let engine = SyncEngine::new();
+        let mut mock_table = MockTable::new();
+        mock_table.commit(&get_init_commit());
+
+        let mut commits = get_segment(&engine, mock_table.table_root(), 0, None)
+            .unwrap()
+            .into_iter();
+
+        let mut scanner = get_commit_log_scanner(&engine, commits.next().unwrap());
+
+        scanner.phase_1().unwrap();
+        assert!(!scanner.has_cdc_action);
+        assert!(scanner.remove_dvs.is_empty());
+        assert!(scanner.add_paths.is_empty());
+
+        scanner.resolve_dvs();
+
+        let x: Vec<_> = scanner
+            .into_scan_batches()
+            .unwrap()
+            .map(|x| -> DeltaResult<_> { Ok(x?.1) })
```

Comment:
`map_ok`?

Reply:
Done. Would like a second pair of eyes on the map_ok+ flatten_ok 🙏

Reply:
LGTM!